### PR TITLE
feat: Adds the ability to update conversation timeframe automatically

### PR DESCRIPTION
**Pull request by iamsivin**
# Pull Request Template

Fixes https://github.com/chatwoot/satisfilabs/issues/24

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

https://user-images.githubusercontent.com/64252451/184308403-194cb9e1-1940-456a-919a-86a7e253b7e1.mov

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      timer: null,
+    };
+  },
+  computed: {
+    hasMinutesDiff() {
+      const today = new Date();
+      const date = new Date(this.datetime * 1000);
+      const diff = Math.abs(today - date);
+      const minutes = Math.floor(diff / 1000 / 60);
+      return minutes <= 60;
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:34]: Use date fns

Comment (iamsivin) [reply app/javascript/dashboard/components/ui/TimeAgo.vue:34]: Done

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      );
+      return years >= 1;
+    },
+    refreshTime() {
+      const ZERO = 0;
+      const MINUTE_IN_MSEC = 60000;
+      const HOUR_IN_MSEC = MINUTE_IN_MSEC * 60;
+      const DAY_IN_MSEC = HOUR_IN_MSEC * 24;
+      const MONTH_IN_MSEC = DAY_IN_MSEC * 30;
+      const YEAR_IN_MSEC = MONTH_IN_MSEC * 12;
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:75]: Move these constants to top

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+          return DAY_IN_MSEC;
+        }
+        if (this.hasMonthDiff && !this.hasYearDiff) {
+          return MONTH_IN_MSEC;
+        }
+        if (this.hasYearDiff) {
+          return YEAR_IN_MSEC;
+        }
+      }
+      return ZERO;
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:94]:
```suggestion
      if (!this.refresh) return ZERO;
      if (this.hasMinutesDiff && !this.hasHourDiff) {
      	return MINUTE_IN_MSEC;
      }
      if (this.hasHourDiff && !this.hasDayDiff) {
      	return HOUR_IN_MSEC;
      }
      if (this.hasDayDiff && !this.hasMonthDiff) {
      	return DAY_IN_MSEC;
      }
      if (this.hasMonthDiff && !this.hasYearDiff) {
      	return MONTH_IN_MSEC;
      }
      if (this.hasYearDiff) {
      	return YEAR_IN_MSEC;
      }
```

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    },
+    hasYearDiff() {
+      const years = differenceInCalendarYears(
+        this.currentDate,
+        this.previousDate
+      );
+      return years >= 1;
+    },
+    refreshTime() {
+      if (this.refresh === true) {
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:78]:
```suggestion
      if (this.refresh) {
```

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      if (timeDiff > DAY_IN_MILLI_SECOND && timeDiff < MONTH_IN_MILLI_SECOND) {
+        return DAY_IN_MILLI_SECOND;
+      }
+      if (timeDiff > MONTH_IN_MILLI_SECOND && timeDiff < YEAR_IN_MILLI_SECOND) {
+        return MONTH_IN_MILLI_SECOND;
+      }
+      if (timeDiff > YEAR_IN_MILLI_SECOND) {
+        return YEAR_IN_MILLI_SECOND;
+      }
+      return ZERO;
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:61]:
```suggestion
      if (timeDiff > YEAR_IN_MSEC) {
        return ZERO;
      }
      if (timeDiff > MONTH_IN_MSEC) {
        return ZERO;
      }
      if (timeDiff > DAY_IN_MSEC) {
        return DAY_IN_MSEC;
      }
      if (timeDiff > HOUR_IN_MSEC) {
        return HOUR_IN_MSEC;
      }
      if (timeDiff > MINUTE_IN_MSEC) {
        return MINUTE_IN_MSEC;
      }

      return ZERO;
```
@iamsivin Please test this suggested change working or not?

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 6]
```diff
@@ -0,0 +1,99 @@
+<template>
+  <span class="time-ago">
+    <span> {{ timeAgo }}</span>
+  </span>
+</template>
+
+<script>
+const ZERO = 0;
+const MINUTE_IN_MILLI_SECOND = 60000;
```
Comment (muhsin-k) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:9]:
```suggestion
const MINUTE_IN_MILLI_SECONDS = 60000;
```

File: app/javascript/dashboard/components/ui/TimeAgo.vue [hunk 7]
```diff
@@ -36,86 +30,55 @@ export default {
...
+      }
+      if (timeDiff > DAY_IN_MILLI_SECOND && timeDiff < MONTH_IN_MILLI_SECOND) {
+        return DAY_IN_MILLI_SECOND;
+      }
+      if (timeDiff > MONTH_IN_MILLI_SECOND && timeDiff < YEAR_IN_MILLI_SECOND) {
+        return MONTH_IN_MILLI_SECOND;
+      }
+      if (timeDiff > YEAR_IN_MILLI_SECOND) {
+        return YEAR_IN_MILLI_SECOND;
       }
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ui/TimeAgo.vue:60]: Please write the comparison in reverse order.

time > year -> return year seconds
time > month -> return month seconds

This way, you wouldn't need to check for `time > hour && time < day.`

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please see the individual comment.

File: app/javascript/dashboard/components/ui/TimeAgo.vue (hunk 7 above)
Comment (iamsivin) [reply app/javascript/dashboard/components/ui/TimeAgo.vue]: Done https://github.com/chatwoot/chatwoot/pull/5253/commits/d9cf0b645453499a7d6cc0ea4cfee402f48f5420

File: app/javascript/dashboard/components/ui/TimeAgo.vue (hunk 5 above)
Comment (iamsivin) [reply app/javascript/dashboard/components/ui/TimeAgo.vue]: Done https://github.com/chatwoot/chatwoot/pull/5253/commits/d9cf0b645453499a7d6cc0ea4cfee402f48f5420

File: app/javascript/dashboard/components/ui/TimeAgo.vue (hunk 7 above)
Comment (muhsin-k) [reply app/javascript/dashboard/components/ui/TimeAgo.vue]: @iamsivin Do we need to set the time interval if the time difference is greater than month/year?
cc @pranavrajs

Comment (iamsivin) [reply app/javascript/dashboard/components/ui/TimeAgo.vue]: @pranavrajs Do we need to do that?

**Comment by muhsin-k** [comment]
@iamsivin Please remove the condition if the time difference is greater than month/year

**Comment by iamsivin** [comment]
> @iamsivin Please remove the condition if the time difference is greater than month/year.

Done

**Review by muhsin-k — APPROVED** [review]
Tested locally and time refreshes as expected.

**Pull request merged**